### PR TITLE
Autograd

### PR DESCRIPTION
**Pull request by xmickos**
• Implemented a system for constructing a graph of calculations.
• Correct aggregation of gradients for reusable tensors.
• Added: matmul, add, mul, sum, reshape, broadcast_to, transpose.
• The necessary intermediate values are saved in the graph for the reverse pass.

### CODE REVIEW

File: include/ttie/ttie.h [hunk 1]
```diff
@@ -8,320 +8,1017 @@
...
+            } else {
+                std::transform(grad.begin(), grad.end(), grad_.begin(), grad.begin(), std::plus<float>());
+            }
+        }
+    };
+
+    struct AddOp final : public GradFn {
+        AddOp(const std::shared_ptr<TensorImpl>& a, const std::shared_ptr<TensorImpl>& b) : GradFn({a, b}) {}
+
+        void backward(TensorImpl& output) override { // c = a + b, ∂c/∂a = 1, ∂c/∂b = 1;
```
Comment (GoshaSerbin) [inline include/ttie/ttie.h:84]: 🤖💬✨

File: include/ttie/ttie.h [hunk 2]
```diff
@@ -8,320 +8,1017 @@
...
+
+    std::string typestr() const noexcept override { return "BroadcastOp"; }
+
+};
+
+
     template <typename T>
-    static std::string vector_to_string(const std::vector<T> &vec, size_t limit = 5)
-    {
+    static std::string vector_to_string(const std::vector<T> &vec, size_t limit = 5) {
```
Comment (GoshaSerbin) [inline include/ttie/ttie.h:290]: Эта строка здесь была всегда, но раз уж она задета, то скажу. Вместо static здесь лучше использовать inline. По свойствам это будет то же самое, внутреннее связывание, но для header файлов в случае со статик будет создаваться копия на каждую единицу трансляции, что увеличивает размер исполняемого файла.

File: include/ttie/ttie.h [hunk 3]
```diff
@@ -8,320 +8,1017 @@
...
-    {
-        std::vector<size_t> shape;
+    class Tensor final {
+        private:
+            std::shared_ptr<TensorImpl> impl_;
 
-        std::vector<float> data;
-        std::vector<float> grad;
+        public:
+            Tensor(bool requires_grad_=false) : impl_(new TensorImpl) { impl_->requires_grad = requires_grad_; }
```
Comment (GoshaSerbin) [inline include/ttie/ttie.h:312]: Здесь и далее new лучше заменить на std::make_shared. Это будет работать чуточку быстрее, т.к. счетчики ссылок на объект и сам объект будут аллоцированы за один вызов и в одном месте, что увеличит попадание в кеш в том числе.

File: include/ttie/ttie.h [hunk 4]
```diff
@@ -8,320 +8,1017 @@
...
-                os << "(not initialized)";
-                return os;
+            void init(std::initializer_list<float> data_) {
+                if(!empty() && data_.size() != size()) {
+                    throw std::invalid_argument("Can't initialize tensor of size " + std::to_string(size()) + " with data which size is " + std::to_string(data_.size()));
+                }
+                impl_->data = data_;
+            }
+
+            Tensor operator=(std::initializer_list<float> list) {
```
Comment (GoshaSerbin) [inline include/ttie/ttie.h:368]: Почему возврат по значению а не по ссылке?

File: tests/test_main.cpp [hunk 5]
```diff
@@ -276,80 +283,78 @@ TEST(ModelTest, ForwardAndBackwardVSTorch) {
...
+    EXPECT_NEAR(l2_bias_grad[0], 2.0000f, 1e-4f);
 }
 
-
 int main(int argc, char **argv) {
     ::testing::InitGoogleTest(&argc, argv);
     return RUN_ALL_TESTS();
-}
\ No newline at end of file
+}
```
Comment (GoshaSerbin) [inline tests/test_main.cpp:360]: Не хватает тестов для добавленной функциональной части, т.е. AddOp, MulOp, и т.д. Проверка того что граф верно строится. Поскольку нововведений много, то есть вероятность багов.

**Review by GoshaSerbin — CHANGES_REQUESTED** [review]
Выглядит здорово! Комменты указал. Не хватает тестов добавленной функциональной части.

**Comment by GoshaSerbin** [comment]
Не забывайте меня указывать assignees справа. Еще я бы рекомендовал чуть более подробно описывать внесенные изменения (можно на русском) в первую очередь для остальных ребят, чтобы им было понятно что вообще просходит

File: tests/test_main.cpp [hunk 6]
```diff
@@ -1,218 +1,323 @@
...
-    tanh.forward(input, output);
+    relu.forward(input, output);
+
+    for (size_t i = 0; i < output.size(); ++i) {
+        EXPECT_FLOAT_EQ(output[i], std::max(0.0f, input[i]));
+    }
+
+    output.backward();
 
-    EXPECT_NEAR(output.data[0], 0.0f, 1e-5f);        // tanh(0) = 0
```
Comment (GoshaSerbin) [inline tests/test_main.cpp:125]: Верните)

File: include/ttie/ttie.h [hunk 7]
```diff
@@ -8,377 +8,1036 @@
...
+            inputs[0]->accumulate_grad(grad_out.get_grad());
+
+            inputs[0]->backward();
         }
-    }
 
-    std::string to_string() const override { return "Sigmoid()"; }
-};
 
-struct Tanh : Layer
```
Comment (GoshaSerbin) [inline include/ttie/ttie.h:264]: Верните

**Review by GoshaSerbin — CHANGES_REQUESTED** [review]
Удален функционал, который уже был запушен, ждем возвращения
